File: weather.py
#!/usr/bin/env python3
"""Weather forecast reader using the Open-Meteo API (free, no API key needed).

Usage:
    python3 weather.py [城市名稱]   # 即時查詢 (需網路)
    python3 weather.py --demo       # 離線展示模式

Examples:
    python3 weather.py 台北
    python3 weather.py Tokyo
    python3 weather.py London
"""

import sys
import json
import urllib.request
import urllib.parse
from datetime import datetime, timedelta


GEOCODING_URL = "https://geocoding-api.open-meteo.com/v1/search"
FORECAST_URL = "https://api.open-meteo.com/v1/forecast"

WMO_CODES = {
    0: "晴天", 1: "大致晴天", 2: "部分多雲", 3: "陰天",
    45: "霧", 48: "結冰霧",
    51: "毛毛雨(輕)", 53: "毛毛雨(中)", 55: "毛毛雨(重)",
    61: "小雨", 63: "中雨", 65: "大雨",
    71: "小雪", 73: "中雪", 75: "大雪", 77: "雪粒",
    80: "陣雨(輕)", 81: "陣雨(中)", 82: "陣雨(強)",
    85: "陣雪(輕)", 86: "陣雪(重)",
    95: "雷陣雨", 96: "雷陣雨夾冰雹", 99: "強雷陣雨夾冰雹",
}

WEEKDAYS = ["一", "二", "三", "四", "五", "六", "日"]


def fetch_json(url: str) -> dict:
    with urllib.request.urlopen(url, timeout=10) as resp:
        return json.loads(resp.read().decode())


def geocode(city: str) -> tuple[float, float, str]:
    params = urllib.parse.urlencode({
        "name": city, "count": 1, "language": "zh", "format": "json"
    })
    data = fetch_json(f"{GEOCODING_URL}?{params}")
    results = data.get("results")
    if not results:
        raise ValueError(f"找不到城市：{city}")
    r = results[0]
    return r["latitude"], r["longitude"], r.get("name", city)


def get_forecast(lat: float, lon: float) -> dict:
    params = urllib.parse.urlencode({
        "latitude": lat,
        "longitude": lon,
        "daily": ",".join([
            "weather_code",
            "temperature_2m_max",
            "temperature_2m_min",
            "precipitation_sum",
            "windspeed_10m_max",
        ]),
        "current": ",".join([
            "temperature_2m",
            "relative_humidity_2m",
            "weather_code",
            "windspeed_10m",
            "apparent_temperature",
        ]),
        "timezone": "auto",
        "forecast_days": 7,
    })
    return fetch_json(f"{FORECAST_URL}?{params}")


def demo_data() -> tuple[str, dict]:
    """Return realistic sample data for offline demonstration."""
    today = datetime.now()
    dates = [(today + timedelta(days=i)).strftime("%Y-%m-%d") for i in range(7)]
    data = {
        "timezone": "Asia/Taipei",
        "current": {
            "temperature_2m": 28.4,
            "apparent_temperature": 32.1,
            "relative_humidity_2m": 78,
            "weather_code": 2,
            "windspeed_10m": 14.5,
        },
        "daily": {
            "time": dates,
            "weather_code": [2, 61, 63, 80, 1, 0, 3],
            "temperature_2m_max": [31.2, 27.5, 25.8, 26.3, 30.1, 32.4, 29.7],
            "temperature_2m_min": [24.6, 22.1, 21.3, 22.0, 23.5, 25.2, 23.8],
            "precipitation_sum": [0.0, 8.2, 15.4, 5.1, 0.0, 0.0, 0.3],
            "windspeed_10m_max": [18.2, 22.7, 28.4, 19.6, 15.3, 12.1, 16.8],
        },
    }
    return "台北 (示範資料)", data


def describe_weather(code: int) -> str:
    return WMO_CODES.get(code, f"代碼{code}")


def display(city_name: str, data: dict) -> None:
    current = data["current"]
    daily = data["daily"]
    tz = data.get("timezone", "")

    print(f"\n{'='*56}")
    print(f"  {city_name} 天氣預報   ({tz})")
    print(f"{'='*56}")

    print(f"\n【目前天氣】")
    print(f"  狀況  : {describe_weather(current['weather_code'])}")
    print(f"  氣溫  : {current['temperature_2m']} °C  (體感 {current['apparent_temperature']} °C)")
    print(f"  濕度  : {current['relative_humidity_2m']} %")
    print(f"  風速  : {current['windspeed_10m']} km/h")

    print(f"\n【未來 7 天預報】")
    header = f"  {'日期':<13} {'天氣':<10} {'最高':>6} {'最低':>6} {'降雨':>8} {'風速':>10}"
    print(header)
    print(f"  {'-'*57}")

    for i, date_str in enumerate(daily["time"]):
        date = datetime.fromisoformat(date_str)
        label = "今天" if i == 0 else f"週{WEEKDAYS[date.weekday()]}"
        desc = describe_weather(daily["weather_code"][i])
        t_max = daily["temperature_2m_max"][i]
        t_min = daily["temperature_2m_min"][i]
        rain = daily["precipitation_sum"][i]
        wind = daily["windspeed_10m_max"][i]
        print(
            f"  {date_str} ({label})  {desc:<8} "
            f"{t_max:>5.1f}°C {t_min:>5.1f}°C "
            f"{rain:>6.1f}mm {wind:>7.1f}km/h"
        )

    print()


def main() -> None:
    args = sys.argv[1:]

    if not args or args[0] == "--demo":
        if not args:
            print("提示：未指定城市，執行離線展示模式。")
            print("用法：python3 weather.py <城市名稱>  (需網路連線)")
            print("      python3 weather.py --demo       (離線展示)\n")
        city_name, data = demo_data()
        display(city_name, data)
        return

    city = args[0]
    print(f"正在查詢「{city}」的天氣預報…")
    try:
        lat, lon, name = geocode(city)
        data = get_forecast(lat, lon)
        display(name, data)
    except ValueError as e:
        print(f"錯誤：{e}", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"網路或解析錯誤：{e}", file=sys.stderr)
        print("提示：可使用 --demo 參數執行離線展示模式", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: weather.js
#!/usr/bin/env node
/**
 * 天氣預報查詢工具 — 使用 Open-Meteo API（免費，無需 API Key）
 *
 * 用法：
 *   node weather.js [城市名稱]   # 即時查詢（需網路）
 *   node weather.js --demo       # 離線展示模式
 *
 * 範例：
 *   node weather.js 台北
 *   node weather.js Tokyo
 *   node weather.js London
 */

const GEOCODING_URL = "https://geocoding-api.open-meteo.com/v1/search";
const FORECAST_URL  = "https://api.open-meteo.com/v1/forecast";

const WMO_CODES = {
  0: "晴天",       1: "大致晴天",    2: "部分多雲",    3: "陰天",
  45: "霧",        48: "結冰霧",
  51: "毛毛雨(輕)", 53: "毛毛雨(中)", 55: "毛毛雨(重)",
  61: "小雨",      63: "中雨",       65: "大雨",
  71: "小雪",      73: "中雪",       75: "大雪",       77: "雪粒",
  80: "陣雨(輕)",  81: "陣雨(中)",   82: "陣雨(強)",
  85: "陣雪(輕)",  86: "陣雪(重)",
  95: "雷陣雨",    96: "雷陣雨夾冰雹", 99: "強雷陣雨夾冰雹",
};

const WEEKDAYS = ["一", "二", "三", "四", "五", "六", "日"];

// ── Helpers ───────────────────────────────────────────────────────────────────

async function fetchJSON(url) {
  const res = await fetch(url, { signal: AbortSignal.timeout(10_000) });
  if (!res.ok) throw new Error(`HTTP ${res.status} ${res.statusText}`);
  return res.json();
}

function describeWeather(code) {
  return WMO_CODES[code] ?? `代碼${code}`;
}

/** Right-pad a string to `len` terminal columns (CJK chars count as 2). */
function padEnd(str, len) {
  let cols = 0;
  for (const ch of str) cols += ch.codePointAt(0) > 0x7f ? 2 : 1;
  return str + " ".repeat(Math.max(0, len - cols));
}

function rpad(n, width, decimals = 1) {
  return n.toFixed(decimals).padStart(width);
}

// ── API calls ─────────────────────────────────────────────────────────────────

async function geocode(city) {
  const url = new URL(GEOCODING_URL);
  url.searchParams.set("name", city);
  url.searchParams.set("count", "1");
  url.searchParams.set("language", "zh");
  url.searchParams.set("format", "json");

  const data = await fetchJSON(url.toString());
  const results = data.results;
  if (!results || results.length === 0) throw new Error(`找不到城市：${city}`);
  const r = results[0];
  return { lat: r.latitude, lon: r.longitude, name: r.name ?? city };
}

async function getForecast(lat, lon) {
  const url = new URL(FORECAST_URL);
  const daily = [
    "weather_code",
    "temperature_2m_max",
    "temperature_2m_min",
    "precipitation_sum",
    "windspeed_10m_max",
  ];
  const current = [
    "temperature_2m",
    "relative_humidity_2m",
    "weather_code",
    "windspeed_10m",
    "apparent_temperature",
  ];
  url.searchParams.set("latitude",     lat);
  url.searchParams.set("longitude",    lon);
  url.searchParams.set("daily",        daily.join(","));
  url.searchParams.set("current",      current.join(","));
  url.searchParams.set("timezone",     "auto");
  url.searchParams.set("forecast_days","7");
  return fetchJSON(url.toString());
}

// ── Demo data ─────────────────────────────────────────────────────────────────

function demoData() {
  const today = new Date();
  const dates = Array.from({ length: 7 }, (_, i) => {
    const d = new Date(today);
    d.setDate(d.getDate() + i);
    return d.toISOString().slice(0, 10);
  });
  return {
    cityName: "台北 (示範資料)",
    data: {
      timezone: "Asia/Taipei",
      current: {
        temperature_2m:       28.4,
        apparent_temperature: 32.1,
        relative_humidity_2m: 78,
        weather_code:         2,
        windspeed_10m:        14.5,
      },
      daily: {
        time:                dates,
        weather_code:        [2, 61, 63, 80, 1, 0, 3],
        temperature_2m_max:  [31.2, 27.5, 25.8, 26.3, 30.1, 32.4, 29.7],
        temperature_2m_min:  [24.6, 22.1, 21.3, 22.0, 23.5, 25.2, 23.8],
        precipitation_sum:   [0.0,  8.2,  15.4,  5.1,  0.0,  0.0,  0.3],
        windspeed_10m_max:   [18.2, 22.7, 28.4, 19.6, 15.3, 12.1, 16.8],
      },
    },
  };
}

// ── Display ───────────────────────────────────────────────────────────────────

function display(cityName, data) {
  const { current, daily } = data;
  const tz = data.timezone ?? "";

  console.log();
  console.log("=".repeat(56));
  console.log(`  ${cityName} 天氣預報   (${tz})`);
  console.log("=".repeat(56));

  console.log("\n【目前天氣】");
  console.log(`  狀況  : ${describeWeather(current.weather_code)}`);
  console.log(`  氣溫  : ${current.temperature_2m} °C  (體感 ${current.apparent_temperature} °C)`);
  console.log(`  濕度  : ${current.relative_humidity_2m} %`);
  console.log(`  風速  : ${current.windspeed_10m} km/h`);

  console.log("\n【未來 7 天預報】");
  console.log(`  ${"日期".padEnd(11)}  ${"天氣".padEnd(8)}   最高   最低     降雨       風速`);
  console.log("  " + "-".repeat(57));

  daily.time.forEach((dateStr, i) => {
    const date  = new Date(dateStr);
    const label = i === 0 ? "今天" : `週${WEEKDAYS[date.getDay() === 0 ? 6 : date.getDay() - 1]}`;
    const desc  = describeWeather(daily.weather_code[i]);
    const tMax  = daily.temperature_2m_max[i];
    const tMin  = daily.temperature_2m_min[i];
    const rain  = daily.precipitation_sum[i];
    const wind  = daily.windspeed_10m_max[i];

    const descCol  = padEnd(desc, 10);
    console.log(
      `  ${dateStr} (${label})  ${descCol}` +
      `${rpad(tMax, 5)}°C ${rpad(tMin, 5)}°C` +
      `${rpad(rain, 7)}mm ${rpad(wind, 8)}km/h`
    );
  });

  console.log();
}

// ── Main ──────────────────────────────────────────────────────────────────────

async function main() {
  const args = process.argv.slice(2);

  if (args.length === 0 || args[0] === "--demo") {
    if (args.length === 0) {
      console.log("提示：未指定城市，執行離線展示模式。");
      console.log("用法：node weather.js <城市名稱>  (需網路連線)");
      console.log("      node weather.js --demo       (離線展示)\n");
    }
    const { cityName, data } = demoData();
    display(cityName, data);
    return;
  }

  const city = args[0];
  console.log(`正在查詢「${city}」的天氣預報…`);
  try {
    const { lat, lon, name } = await geocode(city);
    const data = await getForecast(lat, lon);
    display(name, data);
  } catch (err) {
    if (err.message.startsWith("找不到城市")) {
      console.error(`錯誤：${err.message}`);
    } else {
      console.error(`網路或解析錯誤：${err.message}`);
      console.error("提示：可使用 --demo 參數執行離線展示模式");
    }
    process.exit(1);
  }
}

main();
